apps/src/dance/DanceAiModal.tsx

```diff
@@ -0,0 +1,321 @@
+
+  console.log('do AI:', input);
+
+  const response = await HttpClient.post(
+    CHAT_COMPLETION_URL,
+    JSON.stringify({messages}),
+    true,
+    {
+      'Content-Type': 'application/json; charset=UTF-8',
+    }
+  );
+
+  if (response.status === 200) {
+    const res = await response.json();
+    return res.content;
+  } else {
+    return null;
+  }
+};
+
+interface Item {
+  id: string;
+  name: string;
+}
+
+interface ReturnedItem {
+  id: string;
+  name: string;
+  url: string;
+}
+
+interface DanceAiProps {
+  onClose: () => void;
+}
+
+const DanceAiModal: React.FunctionComponent<DanceAiProps> = ({onClose}) => {
+  const dispatch = useAppDispatch();
+
+  const SLOT_COUNT = 3;
+
+  const inputLibraryFilename = 'ai-inputs';
+  const inputLibrary = require(`@cdo/static/dance/${inputLibraryFilename}.json`);
+
+  const [mode, setMode] = useState('selectInputs');
+  const [currentInputSlot, setCurrentInputSlot] = useState(0);
+  const [inputs, setInputs] = useState<string[]>([]);
+  const [responseParams, setResponseParams] = useState<string>('');
+  const [responseExplanation, setResponseExplanation] = useState<string>('');
+  const [typingDone, setTypingDone] = useState<boolean>(false);
+
+  const showingAi = useSelector((state: {dance: any}) => state.dance.showingAi);
+
+  const getImageUrl = (id: string) => {
+    return `/blockly/media/dance/ai/${id}.png`;
+  };
+
+  const getAllItems = (slotIndex: number) => {
+    if (slotIndex >= SLOT_COUNT) {
+      return [];
+    }
+
+    return inputLibrary.options[slotIndex].map((option: string) => {
+      const item = inputLibrary.items.find((item: Item) => item.id === option);
+      return {
+        id: item.id,
+        name: item.name,
+        url: getImageUrl(item.id),
+      };
+    });
+  };
+
+  const handleItemClick = (id: string) => {
+    if (currentInputSlot < SLOT_COUNT) {
+      setInputs([...inputs, id]);
+      setCurrentInputSlot(currentInputSlot + 1);
+    }
+  };
+
+  const handleGeneratingClick = () => {
+    setMode('results');
+  };
+
+  const handleGenerateClick = () => {
+    const inputNames = inputs.map(
+      input => inputLibrary.items.find((item: Item) => item.id === input).name
+    );
+    const request = `${promptString} ${inputNames.join(', ')}.`;
+    startAi(request);
+    setMode('generating');
+  };
+
+  const startAi = async (value: string) => {
+    const response = await doAi(value);
+
+    const params = JSON.parse(response);
+
+    showingAi.setValue(response);
+
+    const picked = (({
+      backgroundEffect,
+      backgroundColor,
+      foregroundEffect,
+    }) => ({backgroundEffect, backgroundColor, foregroundEffect}))(params);
+
+    const pickedString = JSON.stringify(picked)
+      .replace(/":"/g, '": "')
+      .replace(/","/g, '", "');
+    setResponseParams(`ai(${pickedString})`);
+    setResponseExplanation(params.explanation);
+  };
+
+  const handleDoneClick = () => {
+    dispatch(setShowingAi(undefined));
+  };
+
+  return (
```

Comment:
This `render` function has a lot of complex logic. Knowing that we're under a very tight timeline, maybe later on we can break out into subrender functions for easier readability?

Reply:
Absolutely!  This is still very much a prototype UI, but once we settle on the design, we should decompose this into a collection of functional React components.

Reply:
Sounds good! And this is more of a question for @samantha-code for the next iteration of the modal -  do we want the user to be able to click out of the modal at any point and exit early from the ai generate process?